pkg/ippoolmanager/ippool_informer.go

```diff
@@ -157,22 +192,233 @@ func (im *ipPoolManager) updateSpiderIPPool(ctx context.Context, oldSpiderIPPool
+			// The IPPool resource may no longer exist, in which case we stop
+			// processing.
+			if apierrors.IsNotFound(err) {
+				c.workQueue.Forget(obj)
+				informerLogger.Sugar().Errorf("IPPool '%s' in work queue no longer exists", poolName)
+				return nil
+			}
+
+			c.workQueue.AddRateLimited(poolName)
+			return fmt.Errorf("error syncing '%s': %s, requeuing", poolName, err.Error())
+		}
+
+		err = c.checkScaleIPPool(context.TODO(), pool)
+		if nil != err {
+			// discard some wrong input items
+			if errors.Is(err, constant.ErrWrongInput) {
+				c.workQueue.Forget(obj)
+				informerLogger.Sugar().Errorf("failed to scale IPPool '%s', error: %v", pool.Name, err)
+				return nil
+			}
+
+			// discard some over max retries data
+			if c.workQueue.NumRequeues(obj) < DefaultRetryNum {
+				c.workQueue.AddRateLimited(poolName)
+				return fmt.Errorf("error syncing '%s': %s, requeuing", poolName, err.Error())
+			}
 		}
 
-		informerLogger.Sugar().Infof("update SpiderIPPool '%s' status TotalIPCount to '%d' successfully", currentSpiderIPPool.Name, totalIPCount)
+		c.workQueue.Forget(obj)
+		return nil
+	}
+
+	if err := process(obj); nil != err {
+		informerLogger.Error(err.Error())
+		return true
+	}
+
+	return true
+}
+
+func (c *controller) checkScaleIPPool(ctx context.Context, pool *spiderpoolv1.SpiderIPPool) error {
+	poolLabels := pool.GetLabels()
+	subnetName, ok := poolLabels[constant.LabelIPPoolOwnerSpiderSubnet]
+	if !ok {
+		return fmt.Errorf("%w: there's no owner SpiderSubnet for IPPool '%s'", constant.ErrWrongInput, pool.Name)
+	}
+
+	if pool.Status.AutoDesiredIPCount == nil {
+		informerLogger.Sugar().Debugf("maybe IPPool '%s' is just created for a while, wait for updating status DesiredIPCount", pool.Name)
+		return nil
+	}
+
+	totalIPs, err := spiderpoolip.ParseIPRanges(*pool.Spec.IPVersion, pool.Spec.IPs)
+	if nil != err {
+		return fmt.Errorf("%w: failed to parse IP ranges '%v', error: %v", constant.ErrWrongInput, pool.Spec.IPs, err)
+	}
+
+	desiredIPNum := int(*pool.Status.AutoDesiredIPCount)
+	totalIPCount := len(totalIPs)
+
+	if desiredIPNum == totalIPCount {
+		// no need to scale
+		return nil
+	}
+
+	pool = pool.DeepCopy()
+
+	if desiredIPNum > totalIPCount {
+		// expand
+		ipsFromSubnet, err := c.poolMgr.subnetManager.GenerateIPsFromSubnet(ctx, subnetName, desiredIPNum-totalIPCount)
+		if nil != err {
+			return fmt.Errorf("failed to generate IPs from subnet '%s', error: %v", subnetName, err)
+		}
+
+		informerLogger.Sugar().Infof("try to scale IPPool '%s' IP number from '%d' to '%d' with generated IPs '%v'", pool.Name, totalIPCount, desiredIPNum, ipsFromSubnet)
+		err = c.poolMgr.ScaleIPPoolWithIPs(ctx, pool, ipsFromSubnet, ippoolmanagertypes.Expansion, desiredIPNum)
+		if nil != err {
+			return fmt.Errorf("failed to expand IPPool '%s' with IPs '%v', error: %v", pool.Name, ipsFromSubnet, err)
+		}
+	} else {
+		// shrink: free IP number >= return IP Num
```

Comment:
comment：when it needs to scale down IP,   enough IP is released to make sure it scale down successfully